st2common/st2common/util/config_loader.py

```diff
@@ -110,22 +110,42 @@ def _assign_dynamic_config_values(self, schema, config, parent_keys=None):
         """
         parent_keys = parent_keys or []
 
-        for config_item_key, config_item_value in six.iteritems(config):
-            schema_item = schema.get(config_item_key, {})
+        config_is_dict = isinstance(config, dict)
+        config_is_list = isinstance(config, list)
+        iterator = six.iteritems(config) if config_is_dict else enumerate(config)
+
+        # config_item_key - if config_is_dict then this is the key in the dictionary
+        #                   if config_is_list then this is the index of them item
+        # config_item_value - the value of the key/index for the current item
+        for config_item_key, config_item_value in iterator:
+            if config_is_dict:
+                # different schema for each key/value pair
+                schema_item = schema.get(config_item_key, {})
+            if config_is_list:
+                # same schema is shared between every item in the list
+                schema_item = schema
```

Comment:
Does this work as expected?

For arrays / lists you can specify optional schema for list items using `items` schema property (e.g. https://spacetelescope.github.io/understanding-json-schema/reference/array.html).

If I'm not mistaken, we should probably use `schema.get('items', {})` or similar here, but would need to double check and verify it.

Reply:
Edit: Please ignore my comment.

Just now I see it's correctly handled a couple of lines below (https://github.com/StackStorm/st2/pull/4122/files#diff-1b5471e6a12ebd2f0617b57f18a3c425R140) :)